httpcache.go

```diff
@@ -51,7 +55,48 @@ func CachedResponse(c Cache, req *http.Request) (resp *http.Response, err error)
 	}
 
 	b := bytes.NewBuffer(cachedVal)
-	return http.ReadResponse(bufio.NewReader(b), req)
+	returnResponse, err := http.ReadResponse(bufio.NewReader(b), req)
+	if err != nil {
+		return nil, fmt.Errorf("error loading response from cache: %s\n", err.Error())
+	}
+
+	rangeRaw := req.Header.Get("range")
+	if rangeRaw != "" {
+		tmp := strings.Split(rangeRaw, rangeTypeSeparator)
+		// standard format is bytes=START-END
+		rangetype, rangevalue := tmp[0], tmp[1]
+		if rangetype != "bytes" {
+			fmt.Printf("range type %s not supported\n", rangetype)
+			return returnResponse, nil
+		}
+		// we need to read all the body now, close it, and replace it with another reader
+		// as there is currently no way of "resetting" a Body
+		body, err := ioutil.ReadAll(returnResponse.Body)
+		if err != nil {
+			fmt.Printf("error reading cached response body: %s\n", err.Error())
+			return returnResponse, nil
+		}
+		returnResponse.Body.Close()
+		var rangedRequestStart, rangedRequestEnd int64
+		//TODO(uovobw): handle corrupted/nonstandard request header
+		rangeList := strings.Split(rangevalue, rangeSeparator)
+		// the range is in the form -VAL , the wanted range is (end-val)->end
+		if strings.HasPrefix(rangevalue, rangeSeparator) {
+			rangedRequestEnd = int64(len(body))
+			end, _ := strconv.ParseInt(rangeList[1], 10, 64)
+			rangedRequestStart = rangedRequestEnd - end
+			// the rang is in the form VAL-, the wanted range is val->end
+		} else if strings.HasSuffix(rangevalue, rangeSeparator) {
+			rangedRequestStart, _ = strconv.ParseInt(rangeList[0], 10, 64)
+			rangedRequestEnd = int64(len(body))
+			// normal case with START-END
+		} else {
+			rangedRequestStart, _ = strconv.ParseInt(rangeList[0], 10, 64)
+			rangedRequestEnd, _ = strconv.ParseInt(rangeList[1], 10, 64)
+		}
+		returnResponse.Body = ioutil.NopCloser(bytes.NewReader(body[rangedRequestStart:rangedRequestEnd]))
```

Comment:
This makes it so that the cache stores the requested range. But let's say the first request is for range 20-50 and the next request is for range 0-10. Would the second response's body be bytes 20-30 in the underlying resource (incorrect, it should not take the byte range relative to whatever happens to be cached), or bytes 0-10 (correct)?